src/compiler/emitter.ts

```diff
@@ -2335,46 +2307,130 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
+             * Emitting this as (SubExpr) is really not desirable. We would like to emit the subexpr as is.
+             * Omitting the parentheses, however, could cause change in the semantics of the generated
+             * code if the casted expression has a lower precedence than the rest of the expression, e.g.:
+             *     - (<any>new A).foo should be emitted as (new A).foo and not new A.foo
+             *     - (<any>typeof A).toString() should be emitted as (typeof A).toString() and not typeof A.toString()
+             *     - new (<any>A()) should be emitted as new (A()) and not new A()
+             *     - (<any>function foo() { })() should be emitted as an IIFE (function foo(){})() and not declaration function foo(){} ()
+             *
+             * @param innerExpr The expression found after skipping through nested type assertions and parentheses
+             * @param encounteredParentheses Whether the expression had any pairs of parentheses. We are more conservative in this case.
+             * @param originalParent The parent of the original ParenthesizedExpression or TypeAssertion.
+             */
+            function assertionOrParenExprNeedsToHaveParentheses(innerExpr: Expression, encounteredParentheses: boolean, originalParent: Node): boolean {
+                const originalParentKind = originalParent.kind;
+
+                switch (innerExpr.kind) {
+                    case SyntaxKind.PrefixUnaryExpression:
+                    case SyntaxKind.PostfixUnaryExpression:
+                    case SyntaxKind.VoidExpression:
+                    case SyntaxKind.TypeOfExpression:
+                    case SyntaxKind.DeleteExpression:
+                        return true;
+                    case SyntaxKind.NewExpression:
+                        return encounteredParentheses;
+                    case SyntaxKind.CallExpression:
+                        return originalParentKind === SyntaxKind.NewExpression;
+                    case SyntaxKind.FunctionExpression:
+                        return originalParentKind === SyntaxKind.CallExpression;
+                    case SyntaxKind.NumericLiteral:
+                        return originalParentKind === SyntaxKind.PropertyAccessExpression;
+                }
+
+                const isBinaryExpr = innerExpr.kind === SyntaxKind.BinaryExpression;
+                if (isBinaryExpr) {
+                    const operator = (<BinaryExpression>innerExpr).operatorToken;
+                    if (operator.kind === SyntaxKind.CommaToken) {
+                        return nodeIsSensitiveToCommaExpressions(originalParent);
+                    }
+
+                    if (comparePrecedenceToBinaryPlus(innerExpr) !== Comparison.GreaterThan) {
+                        let parentIsAddition = originalParentKind === SyntaxKind.BinaryExpression &&
+                            (<BinaryExpression>originalParent).operatorToken.kind === SyntaxKind.PlusToken;

```

Comment:
This check is insufficient, but we actually don't have tests for this right now. I'll add some soon.